### PR TITLE
Switch AsyncCollectiveTensor to be a wrapper subclass.

### PR DESCRIPTION
**Pull request by kumpera**
Our usage is of a wrapper, so it makes sense that we use one.

This makes it possible for FakeTensorMode to work.

### CODE REVIEW

File: torch/distributed/_functional_collectives.py [hunk 1]
```diff
@@ -99,43 +98,39 @@ def _wait_tensor(tensor: torch.Tensor) -> torch.Tensor:
         _wait_and_clear_tensor(data_ptr, version_and_work[0])
     return tensor
 
-
 class AsyncCollectiveTensor(torch.Tensor):
-    r"""
```
Comment (wanchaol) [inline torch/distributed/_functional_collectives.py:104]: any reason we delete the comment?

File: torch/distributed/_functional_collectives.py [hunk 2]
```diff
@@ -99,43 +98,39 @@ def _wait_tensor(tensor: torch.Tensor) -> torch.Tensor:
...
-        t = tensor
-        r = torch.Tensor._make_subclass(cls, t, require_grad=t.requires_grad)
-        r._tensor = tensor  # type: ignore[attr-defined]
+    def __new__(cls, elem: torch.Tensor):
+
+        r = torch.Tensor._make_wrapper_subclass(  # type: ignore[attr-defined]
+            cls, elem.size(),
+            strides=elem.stride(), storage_offset=elem.storage_offset(),
+            dtype=elem.dtype, layout=elem.layout,
+            device=elem.device, requires_grad=False
```
Comment (wanchaol) [inline torch/distributed/_functional_collectives.py:115]: should the requires_grad be the same as `elem` too? or we want the wrapper class to not recording the gradient?

**Review by wanchaol — APPROVED** [review]
lgtm, one comment about requires_grad, we might need to carefully set this field so that the behavior is correct.

File: torch/distributed/_functional_collectives.py (hunk 2 above)
Comment (kumpera) [reply torch/distributed/_functional_collectives.py:115]: Grads and wrappers is confusing.
We want only one of the two to support grad and since we don't rewrap the tensor, we should track grads through the wrapped tensor and not the wrapper.

**Review by wconstab — APPROVED** [review]
nice! cc @animesh does this also help unblock dynamo tracing, or does dynamo also not support wrapper subclasses currently?
